File: src/common/py/data/entities/project.py
import dataclasses
import typing
from dataclasses import dataclass, field
from enum import IntEnum

from dataclasses_json import dataclass_json

from .project_feature import ProjectFeature, ProjectFeatureID

ProjectID = int


@dataclass_json
@dataclass(frozen=True, kw_only=True)
class Project:
    """
    Data for a single **Project**.

    Attributes:
        project_id: The unique project identifier.
        creation_time: A UNIX timestamp of the project creation time.
        title: The title of the project.
        description: An optional project description.
        status: The project status.
        features: The data of the various project features.
        features_selection: List of enabled user-selectable features.
    """

    class Status(IntEnum):
        """
        The status of a project.
        """

        ACTIVE = 0x0
        DELETED = 0xFF

    project_id: ProjectID

    creation_time: float

    title: str
    description: str

    status: Status = Status.ACTIVE

    features: typing.Dict[ProjectFeatureID, ProjectFeature] = dataclasses.field(
        default_factory=dict
    )
    features_selection: typing.List[ProjectFeatureID] = dataclasses.field(
        default_factory=list
    )

File: src/common/web/data/entities/Project.ts
import { Type } from "class-transformer";

import { type ProjectID } from "./EntityTypes";
import { ProjectFeatures } from "./features/ProjectFeatures";

/*
 * The status of a project.
 */
export const enum ProjectStatus {
    Active = 0x0,
    Deleted = 0xff,
}

/**
 * Data for a single **Project**.
 *
 * @param project_id - The unique project identifier.
 * @param creation_time - A UNIX timestamp of the project creation time.
 * @param title - The title of the project.
 * @param description - An optional project description.
 * @param status - The project status.
 * @param features - All project features data.
 */
export class Project {
    public readonly project_id: ProjectID;

    public readonly creation_time: number;

    public readonly title: string;
    public readonly description: string;

    public readonly status: ProjectStatus = ProjectStatus.Active;

    // @ts-ignore
    @Type(() => ProjectFeatures)
    public readonly features: ProjectFeatures = new ProjectFeatures();

    public constructor(projectID: ProjectID, creationTime: number, title: string, description: string = "", features: ProjectFeatures = new ProjectFeatures()) {
        this.project_id = projectID;

        this.creation_time = creationTime;

        this.title = title;
        this.description = description;

        this.features = features;
    }
}
